src/bootstrap/src/lib.rs

```diff
@@ -1056,11 +1056,29 @@ Executed at: {executed_at}"#,
             }
         };
 
-        let fail = |message: &str| {
+        let fail = |message: &str, output: CommandOutput| -> ! {
             if self.is_verbose() {
                 println!("{message}");
             } else {
-                println!("Command has failed. Rerun with -v to see more details.");
+                let (stdout, stderr) = (output.stdout_if_present(), output.stderr_if_present());
+                // If the command captures output, the user would not see any indication that
+                // it has failed. In this case, print a more verbose error, since to provide more
+                // context.
+                if stdout.is_some() || stderr.is_some() {
+                    if let Some(stdout) = output.stdout_if_present() {
+                        if !stdout.trim().is_empty() {
+                            println!("STDOUT:\n{stdout}\n");
+                        }
+                    }
+                    if let Some(stderr) = output.stderr_if_present() {
+                        if !stderr.trim().is_empty() {
+                            println!("STDERR:\n{stderr}\n");
+                        }
+                    }
+                    println!("Command {command:?} has failed. Rerun with -v to see more details.");
+                } else {
+                    println!("Command has failed. Rerun with -v to see more details.");
```

Comment:
Isn't too verbose to print `BootstrapCommand` in debug mode?

Reply:
I'm not sure. Seeing stdout/stderr helps, but I'm always a bit frustrated when bootstrap says that a command has failed, but does not tell me *which* command it was. We could in theory print e.g. just the name of the executed binary, but the whole command line usually isn't that long, and provides useful information (e.g. you can try to copy-paste it and run it yourself).

Reply:
In that case, shouldn't we print it unconditionally?

Reply:
I also thought about that, but I think there is still a distinction between captured and uncaptured commands. When I use the example with `x perf <wrong-arg>`, the output will be a help message from clap; I don't want to pollute that with the executed command.

When we run an uncaptured command, it is assumed that the stdout/stderr is somehow readable by users, and will help them figure out potential issues. In that case, I think that printing the command is not always needed. When we run a captured command, the user is not supposed to see stdout/stderr, so maybe it's not enough to debug the issue. When such command fails, we try to show both the output and the command, to provide more context.

Reply:
> I also thought about that, but I think there is still a distinction between captured and uncaptured commands. When I use the example with x perf <wrong-arg>, the output will be a help message from clap; I don't want to pollute that with the executed command.

Right, but saying "Command has failed. Rerun with -v to see more details." doesn't seem useful. Instead, it should specify the executed command, e.g., "Command x perf <wrong-arg> failed to execute. Rerun with -v to see more details.". Otherwise we can't know which command was failed, or do we print it somewhere else?

Another thing: If printing `BootstrapCommand` prints the stderr/stdout (based on "When I use the example with x perf <wrong-arg>, the output will be a help message from clap" which was your answer above), why are we also printing stdout and stderr separately? Wouldn't that make it excessively verbose?